drivers/wdm/audio/backpln/portcls/irpstream.cpp

```diff
@@ -779,23 +780,50 @@ CIrpQueue::GetCurrentIrpOffset()
 BOOLEAN
 NTAPI
 CIrpQueue::GetAcquiredTagRange(
-    IN PVOID * FirstTag,
-    IN PVOID * LastTag)
+    OUT PVOID * FirstTag,
+    OUT PVOID * LastTag)
 {
     KIRQL OldLevel;
     BOOLEAN Ret = FALSE;
-    //PIRP Irp;
-    //PLIST_ENTRY CurEntry;
-    //PKSSTREAM_DATA StreamData;
+    PIRP Irp;
+    PLIST_ENTRY CurEntry;
+    PKSSTREAM_DATA StreamData;
 
     // lock list
     KeAcquireSpinLock(&m_IrpListLock, &OldLevel);
 
-    // initialize to zero
-    *FirstTag = NULL;
-    *LastTag = NULL;
+    CurEntry = m_IrpList.Flink;
+
+    while (CurEntry != &m_IrpList)
+    {
+        // get irp from list entry
+        Irp = (PIRP)CONTAINING_RECORD(CurEntry, IRP, Tail.Overlay.ListEntry);
+
+        // get stream data
+        StreamData = (PKSSTREAM_DATA)Irp->Tail.Overlay.DriverContext[STREAM_DATA_OFFSET];
+
+        if (StreamData->StreamHeaderIndex == 0)
+        {
+            // store the first tag from stream header index
+            *FirstTag = StreamData->Tags[StreamData->StreamHeaderIndex].Tag;
+        }
 
-    UNIMPLEMENTED;
+        // increment stream header index
+        StreamData->StreamHeaderIndex++;
+
+        if (StreamData->StreamHeaderIndex == StreamData->StreamHeaderCount)
```

Comment:
You don't need to loop the list if you're only interested in the first and last item.
Though it feels like these should be "smallest" and "largest" more so than first/last in the list? But I don't have a good understanding of what these tags mean, so maybe not.

Reply:
Looked at this more, and the concept of first/last is correct the way you have it.
But I think what you want is the `Tags[0]` on the first IRP, and `Tags[StreamHeaderIndex - 1]` on the last IRP where `StreamHeaderIndex` is not zero. So something like
```
FirstTagSet = FALSE;
while (...)
{
    Irp = ...;
    StreamData = ...;
    if (StreamData->StreamHeaderIndex == 0)
    {
        break;
    }
    if (!FirstTagSet)
    {
        *FirstTag = StreamData->Tags[0].Tag;
        FirstTagSet = TRUE;
    }
    *LastTag = StreamData->Tags[StreamData->StreamHeaderIndex - 1].Tag;
    ...
}
```
Or maybe you have to go to `StreamHeaderCount` and actually do a loop. That part's not clear to me -- `ReleaseMappingWithTag` seems to look up to Index on the current IRP, up to Count on others... not sure what that's about.